malcolm/modules/scanning/parts/scanrunnerpart.py

```diff
@@ -0,0 +1,498 @@
+            generators.append(self.parse_line_generator(generator['line']))
+
+        kwargs_list = ['continuous', 'delay_after']
+        kwargs = self.get_kwargs_from_dict(entry, kwargs_list)
+
+        return CompoundGenerator(generators, duration=duration, **kwargs)
+
+    def parse_scan(self, entry):
+        # type: (dict) -> None
+        name = entry['name']
+        generator = self.parse_compound_generator(entry['generator'])
+        kwargs = self.get_kwargs_from_dict(entry, 'repeats')
+
+        self.scan_sets[name] = Scan(name, generator, **kwargs)
+
+    @staticmethod
+    def get_current_datetime(time_separator=":"):
+        # type: (str) -> str
+        return datetime.now().strftime("%Y-%m-%d-%H{sep}%M{sep}%S".format(
+            sep=time_separator
+        ))
+
+    # noinspection PyPep8Naming
+    def loadFile(self):
+        # type: () -> None
+
+        # Update state
+        self.set_runner_state(RunnerStates.LOADING)
+        self.runner_status_message.set_value("Loading scan file")
+
+        # Read contents of file into string
+        file_contents = self.get_file_contents()
+
+        # Parse the string
+        parsed_yaml = self.parse_yaml(file_contents)
+
+        # Empty the current dictionaries
+        self.scan_sets = {}
+
+        # Parse the configuration
+        for item in parsed_yaml:
+            key_name = list(item.keys())[0].upper()
+            if key_name == EntryType.SCAN.name:
+                self.parse_scan(item['scan'])
+            else:
+                raise ValueError(
+                    "Unidentified object in YAML: {key}".format(key=key_name))
+
+        # Count the number of scans configured
+        self.update_scans_configured()
+
+        self.set_runner_state(RunnerStates.CONFIGURED)
+        self.runner_status_message.set_value("Load complete")
+
+    def update_scans_configured(self):
+        # type: () -> None
+        number_of_scans = 0
+        for key in self.scan_sets:
+            number_of_scans += self.scan_sets[key].repeats
+        self.scans_configured.set_value(number_of_scans)
+
+    def create_directory(self, directory):
+        # type: (str) -> None
+        try:
+            os.mkdir(directory)
+        except OSError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value(
+                "Could not create directory")
+            raise IOError(
+                "ERROR: unable to create directory: {dir}".format(
+                    dir=directory
+                ))
+
+    def create_and_get_sub_directory(self, root_directory):
+        # type: (str) -> str
+        today_str = self.get_current_datetime(time_separator="-")
+        sub_directory = "{root}/{scan_mri}-{date}".format(
+            root=root_directory, scan_mri=self.mri, date=today_str)
+        self.create_directory(sub_directory)
+        return sub_directory
+
+    @add_call_types
+    def run(self, context):
```

Comment:
I take it we don't care about aborting these scans when they are kicked off?

Reply:
I was planning on adding it at some point. What would the best way of handling this be?

Reply:
Store context from run, then make a new function abort() that calls context.stop(). That will catch run() when it is waiting for the scan to run, but not abort the current running scan. To do that you would also have to abort() the last scan that was run.